compiler-core/src/type_/expression.rs

```diff
@@ -18,123 +18,184 @@ use im::hashmap;
-        }
-    }
-
-    pub fn intersect(&self, targets: SupportedTargets) -> SupportedTargets {
-        SupportedTargets {
-            erlang: self.erlang && targets.erlang,
-            javascript: self.javascript && targets.javascript,
-        }
-    }
-
-    pub fn merge(&self, targets: SupportedTargets) -> SupportedTargets {
-        SupportedTargets {
-            erlang: self.erlang || targets.erlang,
-            javascript: self.javascript || targets.javascript,
-        }
-    }
-
-    pub fn add(&self, target: Target) -> SupportedTargets {
-        match target {
-            Target::Erlang => SupportedTargets {
-                erlang: true,
-                javascript: self.javascript,
-            },
-            Target::JavaScript => SupportedTargets {
-                erlang: self.erlang,
-                javascript: true,
-            },
+    /// For values that have anexternal implementation for _all_ gleam targets
+    /// but do not have a default Gleam implementation.
+    ///
+    /// Some examples:
+    ///
+    /// ```gleam
+    /// @external(erlang, "foo", "bar")
+    /// @external(javascript, "foo", "bar")
+    /// pub fn foo() -> Int
+    /// ```
+    pub fn all_externals() -> Implementations {
+        Implementations {
+            gleam: false,
+            erlang: true,
+            javascript: true,
         }
     }
 
+    /// Given the implementations of a function update those with taking into
+    /// account the `implementations` of another function (or constant) used
+    /// inside its body.
+    pub fn update_from_use(&mut self, implementations: &Implementations) {
+        // With this pattern matching we won't forget to deal with new targets
+        // when those are added :)
+        let Implementations {
+            gleam,
+            erlang,
+            javascript,
+        } = implementations;
+
+        // If a pure-Gleam function uses a function that doesn't have a pure
+        // Gleam implementation, then it's no longer pure-Gleam.
+        self.gleam = self.gleam && *gleam;
+        // If a function uses a function that has an erlang specific (or
+        // javascript specific) implementation then it is considered as having
+        // a target specific implementation as well.
+        //
+        // For example:
+        // ```gleam
+        // @external(erlang, "foo", "bar")
+        // pub fn erlang_only_with_pure_gleam_default() -> Int {
+        //   1 + 1
+        // }
+        //
+        // pub fn main() { erlang_only_with_pure_gleam_default() }
+        // ```
+        // Both functions will end up having the following `Implementations`:
+        // `Implementations { gleam: true, erlang: true, javascript: false}`.
+        // They have a pure gleam implementation and an erlang specific external
+        // implementation.
+        self.erlang = self.erlang || *erlang;
+        self.javascript = self.javascript || *javascript;
```

Comment:
I think these should be &&, not ||. If code only supports JavaScript and then calls a function which does support Erlang then it does not start supporting Erlang. The previous code would still mean the function doesn't work on Erlang.

Reply:
Ah I think we're interpreting these booleans in a different way.
To me the field `erlang/javascript` set to `true` means that the function relies somewhere on an external Erlang/JavaScript implementation.

I interpret this `Implementations { gleam: true, erlang: true, javascript: false }` as:
- The function has a pure gleam implementation that can be run on any target
- On the Erlang target this function is going to use external code
- On the JavaScript target this function is not going to use external code

And I interpret the following `Implementation { gleam: false, erlang: true, javascript: true }` as:
- The function doesn't have a gleam implementation that can be run on any target
- On the Erlang target it's going to use external code
- On the JavaScript target it's going to use external code
- So it supports all _current_ targets but not future ones

Does that make sense to you?

Reply:
Ah I see! That seems good to me! Thank you

Maybe we can find names that make this super clear. How about:

```rust
Implementations {
    pure_gleam: bool,
    uses_erlang_externals: bool,
    users_javascript_externals: bool,  
}
```

Reply:
Yeah I think that's perfect, thank you!